File: Python/0_jsonmapping.py
import json

def rule_converter(input: dict, mapping_rules: dict):
    splitter = '.'
    result = {}
    for (input_attribute, output_attribute) in mapping_rules.items():
        value = 'Not Assigned'
        if(splitter in input_attribute):
            input_attributes_chain = input_attribute.split(splitter)
            obj = input
            for attr in input_attributes_chain:
                obj = obj[attr]
            value = obj
        else: 
            value = input[input_attribute]
        print("\nThe input value is : {}".format(value))

        if (splitter in output_attribute):
            output_attribute_chain = output_attribute.split(splitter)
            obj = result
            length = len(output_attribute_chain)
            for i in range(length):
                key = output_attribute_chain[i]
                if key not in obj.keys():
                    obj[key] = value if (i==length-1) else {}
                obj = obj[key]
        else:
            result[output_attribute] = value

        print("input_attribute is {}, output_attribute is {}".format(input_attribute, output_attribute))

    return result

def test(input, mapping_rules, expected):
    print("\nThe input is :\n{}".format(input))
    print("\nThe mapping rule is:\n{}".format(mapping_rules))
    output = rule_converter(input, mapping_rules)
    print("The ouput is : \n{}".format(output))
    print("-------------------------------------------\nTest result is: {} \n -----------------------------------------".format(json.dumps(expected) == json.dumps(output)))

if __name__ == "__main__":
    input = {
        'i_a': {
            'i_a_a': 12,
            'i_a_b': 'value of i_a_b'
        },
        'i_b': "abc",
        'i_c': "value of i_c"
    }
    mapping_rules = {
        'i_a.i_a_a': 'o_a',
        'i_b': 'o_b.o_b_a',
        'i_c': 'o_c'
    }
    expected =     {
        "o_a": 12,
        "o_b": {
        "o_b_a": "abc"
        },
        "o_c": "value of i_c"
    }
    test(input,mapping_rules, expected)

    input = {
        'i_a': {
            'i_a_a': 'value of i_a_a',
            'i_a_b': 'value of i_a_b',
            'i_a_c': 'value of i_a_c',
        },
        'i_b': 'abc',
        'i_c': 'value of i_c',
        'i_d': 'value of i_d',
    }
    mapping_rules = {
        'i_a.i_a_a': 'o_a.o_a_a',
        'i_a.i_a_b': 'o_a.o_a_b',
        'i_a.i_a_c': 'o_a.o_a_c',
        'i_b': 'o_b',
        'i_c': 'o_c',
        'i_d': 'o_d',
    }
    expected =   {
        "o_a": {
        "o_a_a": "value of i_a_a",
        "o_a_b": "value of i_a_b",
        "o_a_c": "value of i_a_c"
        },
        "o_b": "abc",
        "o_c": "value of i_c",
        "o_d": "value of i_d"
    }
    test(input,mapping_rules, expected)

File: Javascript_Typescript/0_jsonmapping.ts
/*jshint esversion: 6 */

function converter(input, mapping_rules){
    let result = {};
    const splitter = '.';
    for(let key in  mapping_rules){
        let input_attribute = key;
        let output_attribute = mapping_rules[key];
        let value = {};

        console.log("The mapping key value is:");
        console.log(input_attribute, output_attribute);
        
        
        if(input_attribute.indexOf(splitter) == -1) {//not found
            console.log("The input attributes: without comma:",input_attribute);
            value = input[input_attribute];
            console.log("The value get is : " + value);
        }
        else{
            console.log("The input attributes: with comma:",input_attribute);
            let input_attributes_chain = input_attribute.split(splitter);
            obj = input;
            for(const attr of input_attributes_chain){
                obj = obj[attr];
            }
            value = obj;
            console.log("The value get is : " + value);
        }

        if(output_attribute.indexOf(splitter) == -1){
            console.log("The output_attributes: without comma:", output_attribute);
            result[output_attribute] = value;
            console.log("The result is : ", result);
        }
        else{
            console.log("The output_attributes: with comma:", output_attribute);
            let output_attribute_chain = output_attribute.split(splitter);
            obj = value;
            for(const attr of output_attribute_chain.reverse()){
                temp = {};
                temp[attr] = obj;
                obj = temp;
            }
            result = Object.assign(obj,result);
            console.log("The result is : ", result);
        }
    }
    return result;
}

input = {
    'i_a': {
        'i_a_a': 12,
        'i_a_b': 'value of i_a_b'
    },
    'i_b': "abc"
};

mapping_rules = {
    'i_a.i_a_a': 'o_a',
    'i_b': 'o_b.o_b_a'
};

console.log(`The input is: \n`);
console.log(input);

console.log("The mapping rule is: \n");
console.log(mapping_rules);

output = converter(input, mapping_rules);
console.log('The output is: \n');
console.log(JSON.stringify(output));